### PR TITLE
[Merged by Bors] - feat(*): some auxiliary definitions for toAdditive

### PR DESCRIPTION
**Pull request by fpvandoorn**
* Define some extra functions about List, String, Name, Binder and Expr (ported from Lean 3)
* Move some definitions about Expr from `Tactic/Core` to `Lean/Expr`.

### CODE REVIEW

File: Mathlib/Lean/Expr.lean [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+namespace Lean
+
+namespace BinderInfo
+
+/-! ### Declarations about `BinderInfo` -/
+
+
+instance : Inhabited BinderInfo :=
+  ⟨BinderInfo.default⟩
```
Comment (gebner) [inline Mathlib/Lean/Expr.lean:25]:
```suggestion
```
already in core

File: Mathlib/Lean/Expr.lean [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+| _ => ("(", ")")
+
+end BinderInfo
+
+/-- The type of binders containing a name, the binder info and the binding type -/
+structure Binder :=
+  (name : Name)
+  (info : BinderInfo)
+  (type : Expr)
+deriving Inhabited
```
Comment (gebner) [inline Mathlib/Lean/Expr.lean:41]:
```suggestion
structure Binder where
  name : Name
  info : BinderInfo
  type : Expr
  deriving Inhabited
```

I think this structure should be deleted completely.  The idiomatic way is to use free variables (mkFVar) and then look up the user name, binder info, type, etc. in the local context.

The Binder structure doesn't do let bindings either.

File: Mathlib/Lean/Expr.lean [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+namespace Name
+
+/-! ### Declarations about `name` -/
+
+/-- Find the largest prefix `n` of a `Name` such that `f n != none`, then replace this prefix
+with the value of `f n`. -/
+def mapPrefix (f : Name → Option Name) : Name → Name
+| anonymous => anonymous
+| str n' s d => (f (mkStr n' s)).getD (mkStr (mapPrefix f n') s)
+| num n' nr d => (f (mkNum n' nr)).getD (mkNum (mapPrefix f n') nr)
```
Comment (gebner) [inline Mathlib/Lean/Expr.lean:72]: Some new syntax you might want to check out:
```suggestion
def mapPrefix (f : Name → Option Name) (n : Name) : Name := do
  if let some n' := f n then return n'
  match n with
  | anonymous => anonymous
  | str n' s _ => mkStr (mapPrefix f n') s
  | num n' i _ => mkNum (mapPrefix f n') i
```

I would strongly avoid giving a name to the `UInt64` argument, you can only do bad things with it.

File: Mathlib/Lean/Expr.lean [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+namespace Expr
+
+private def getAppFnArgsAux : Expr → Array Expr → Nat → Name × Array Expr
+  | app f a _,   as, i => getAppFnArgsAux f (as.set! i a) (i-1)
+  | const n _ _, as, i => (n, as)
+  | _,           as, _ => (Name.anonymous, as)
+
+def getAppFnArgs (e : Expr) : Name × Array Expr :=
+  let nargs := e.getAppNumArgs
+  getAppFnArgsAux e (mkArray nargs arbitrary) (nargs-1)
```
Comment (gebner) [inline Mathlib/Lean/Expr.lean:85]:
```suggestion
  withApp e Prod.mk
```

But really, just use `getAppFn` and `getAppArgs` directly.

File: Mathlib/Lean/Expr.lean [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  This definition doesn't instantiate bound variables, and therefore produces a term that is open.
+  See note [open expressions]. -/
+def piCodomain : Expr → Expr
+| forallE n t b d => piCodomain b
+| e => e
+
+/-- Get the binders and codomain of a pi-type.
+  This definition doesn't instantiate bound variables, and therefore produces a term that is open.
+  See note [open expressions]. -/
+def piBinders : Expr → List Binder × Expr
```
Comment (gebner) [inline Mathlib/Lean/Expr.lean:111]: Please don't.  Just use `forallTelescope` instead.

**Review by gebner — CHANGES_REQUESTED** [review]

**Comment by fpvandoorn** [comment]
Thanks for all the pointers! I'm not used to the idiomatic ways to do things in Lean 4 (and there are relatively few examples to learn from), so these are incredibly helpful! I will remove all the things that have better alternatives.

File: Mathlib/Lean/Expr.lean (hunk 4 above)
Comment (fpvandoorn) [reply Mathlib/Lean/Expr.lean]: The first component of the function is slightly different than `getAppFn`. It is used in `Ring` and `NormNum` already.

File: Mathlib/Data/Array/Defs.lean [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+/-!
+## Definitions on Arrays
+
+This file contains various definitions on `Array`. It does not contain
+proofs about these definitions, those are contained in other files in `Mathlib.Data.Array`.
+-/
+
+variable {α β : Type _}
+
```
Comment (gebner) [inline Mathlib/Data/Array/Defs.lean:17]:
```suggestion
```

These can be auto-bound now.

File: Mathlib/Data/Array/Defs.lean [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+/-- Given a function `f : ℕ → α → β` and `as : Array α`, `as = #[a₀, a₁, ...]`, returns the Array
+`#[f 0 a₀, f 1 a₁, ...]`. -/
+def mapWithIndex (f : ℕ → α → β) (as : Array α) : Array β :=
+(as.zip $ range as.size).map λ p => f p.2 p.1
+
+/-- Monad variant of `mapWithIndex` (this could be generalized to `Applicative`). -/
+def mapWithIndexM {m : Type v → Type w} [Monad m] (f : ℕ → α → m β) (as : Array α) :
+  m (Array β) :=
+(as.zip $ range as.size).mapM λ p => f p.2 p.1
```
Comment (gebner) [inline Mathlib/Data/Array/Defs.lean:32]:
```suggestion
  as.mapIdxM (f ∙ ∙)
```

Or just use mapIdxM directly.

File: Mathlib/Data/Array/Defs.lean [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+namespace Array
+
+/-- The array `#[0, 1, ..., n - 1]`. -/
+def range (n : Nat) : Array ℕ :=
+(List.range n).toArray
+
+/-- Given a function `f : ℕ → α → β` and `as : Array α`, `as = #[a₀, a₁, ...]`, returns the Array
+`#[f 0 a₀, f 1 a₁, ...]`. -/
+def mapWithIndex (f : ℕ → α → β) (as : Array α) : Array β :=
+(as.zip $ range as.size).map λ p => f p.2 p.1
```
Comment (gebner) [inline Mathlib/Data/Array/Defs.lean:27]:
```suggestion
  as.mapIdx (f ∙ ∙)
```
Or maybe just use mapIdx directly.

File: Mathlib/Data/Array/Defs.lean [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+proofs about these definitions, those are contained in other files in `Mathlib.Data.Array`.
+-/
+
+variable {α β : Type _}
+
+namespace Array
+
+/-- The array `#[0, 1, ..., n - 1]`. -/
+def range (n : Nat) : Array ℕ :=
+(List.range n).toArray
```
Comment (gebner) [inline Mathlib/Data/Array/Defs.lean:22]:
```suggestion
def range (n : Nat) : Array ℕ :=
  n.fold (flip Array.push) #[]
```
The list construction is quite inefficient.

File: Mathlib/Data/List/Defs.lean [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+proofs about these definitions, those are contained in other files in `Mathlib.Data.List`.
+-/
+
+variable {α β : Type _}
+
+namespace List
+
+/-- Given a function `f : ℕ → α → β` and `as : list α`, `as = [a₀, a₁, ...]`, returns the list
+`[f 0 a₀, f 1 a₁, ...]`. -/
+def mapWithIndex (f : ℕ → α → β) (as : List α) : List β :=
```
Comment (gebner) [inline Mathlib/Data/List/Defs.lean:22]:
```suggestion
`[f 0 a₀, f 1 a₁, ...]`. -/
def mapIdx (as : List α) (f : ℕ → α → β) : List β :=
```
For consistency with arrays.

File: Mathlib/Tactic/Core.lean [hunk 11]
```diff
@@ -5,27 +5,19 @@ Authors: Mario Carneiro, Aurélien Saue
...
--- syntax simpArg := simpStar <|> simpErase <|> simpLemma
-def simpArg := simpStar.binary `orelse (simpErase.binary `orelse simpLemma)
-
-end Lean.Parser.Tactic
-
-namespace Lean.Expr
+def setProtected {m : Type → Type} [Monad m] [MonadEnv m] (nm : Name) : m Unit := do
+  let env ← getEnv
+  let env := addProtected env nm
+  setEnv env
```
Comment (gebner) [inline Mathlib/Tactic/Core.lean:15]:
```suggestion
  modifyEnv (addProtected ∙ nm)
```

**Comment by fpvandoorn** [comment]
Thanks for the review!

**Pull request closed without merging**